File: 0443-string-compression/0443-string-compression.ts
function compress(chars: string[]): number {
    let compressed = "";
    let count = 1;
    for (let i = 1; i < chars.length; i++) {
        if (chars[i] == chars[i - 1]) {
            count++;
        } else {
            compressed += chars[i - 1];
            if (count > 1) {
                compressed += count;
            }
            count = 1;
        }
    }
  
    //last character count
    compressed += chars[chars.length - 1];
    if (count > 1) {
        compressed += count;
    }
    //modifying the input array
    for (let i = 0; i < compressed.length; i++) {
        chars[i] = compressed[i];
    }
 
    return compressed.length;
};

File: 0443-string-compression/0443-string-compression.java
class Solution {
    public int compress(char[] chars) {
        int N = chars.length;
        int i = 0;
        int j = 0;
        while (i < N) {
            int count = 0;
            char c = chars[i];
            while (i < N && chars[i] == c) {
                i++;
                count++;
            }
            chars[j++] = c;
            if (count > 1) {
                for (char ch : String.valueOf(count).toCharArray()) {
                    chars[j++] = ch;
                }
            }
        }
        return j;
    }
}